Content.Shared/SegmentedEntity/SegmentedEntitySystem.cs

```diff
@@ -0,0 +1,306 @@
+using Content.Shared.Inventory.Events;
+using Content.Shared.Tag;
+using Content.Shared.Storage.Components;
+using Content.Shared.Weapons.Ranged.Events;
+using Content.Shared.Teleportation.Components;
+using Robust.Shared.Map;
+using Robust.Shared.Physics.Systems;
+using Robust.Shared.Physics.Components;
+using System.Numerics;
+using Robust.Shared.Network;
+using Robust.Shared.GameStates;
+
+namespace Content.Shared.SegmentedEntity
+{
+    public sealed partial class LamiaSystem : EntitySystem
+    {
+        [Dependency] private readonly TagSystem _tagSystem = default!;
+        [Dependency] private readonly DamageableSystem _damageableSystem = default!;
+        [Dependency] private readonly SharedJointSystem _jointSystem = default!;
+        [Dependency] private readonly INetManager _net = default!;
+        [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+        Queue<(SegmentedEntitySegmentComponent segment, EntityUid lamia)> _segments = new();
+
+        [ValidatePrototypeId<TagPrototype>]
+        private const string LamiaHardsuitTag = "AllowLamiaHardsuit";
+        public override void Initialize()
+        {
+            base.Initialize();
+            //Parent subscriptions
+            SubscribeLocalEvent<SegmentedEntityComponent, HitScanAfterRayCastEvent>(OnShootHitscan);
+            SubscribeLocalEvent<SegmentedEntityComponent, InsertIntoEntityStorageAttemptEvent>(OnLamiaStorageInsertAttempt);
+            SubscribeLocalEvent<SegmentedEntityComponent, DidEquipEvent>(OnDidEquipEvent);
+            SubscribeLocalEvent<SegmentedEntityComponent, DidUnequipEvent>(OnDidUnequipEvent);
+            SubscribeLocalEvent<SegmentedEntityComponent, ComponentInit>(OnInit);
+            SubscribeLocalEvent<SegmentedEntityComponent, ComponentShutdown>(OnShutdown);
+            SubscribeLocalEvent<SegmentedEntityComponent, JointRemovedEvent>(OnJointRemoved);
+            SubscribeLocalEvent<SegmentedEntityComponent, EntParentChangedMessage>(OnParentChanged);
+            SubscribeLocalEvent<SegmentedEntityComponent, StoreMobInItemContainerAttemptEvent>(OnStoreSnekAttempt);
+
+            //Child subscriptions
+            SubscribeLocalEvent<SegmentedEntitySegmentComponent, InsertIntoEntityStorageAttemptEvent>(OnSegmentStorageInsertAttempt);
+            SubscribeLocalEvent<SegmentedEntitySegmentComponent, GetExplosionResistanceEvent>(OnSnekBoom);
+            SubscribeLocalEvent<SegmentedEntitySegmentComponent, DamageChangedEvent>(HandleDamageTransfer);
+            SubscribeLocalEvent<SegmentedEntitySegmentComponent, DamageModifyEvent>(HandleSegmentDamage);
+        }
+        public override void Update(float frameTime)
+        {
+            //I HATE THIS, SO MUCH. I AM FORCED TO DEAL WITH THIS MONSTROSITY. PLEASE. SEND HELP.
+            base.Update(frameTime);
+            foreach (var segment in _segments)
+            {
+                var segmentUid = segment.segment.Owner;
```

Comment:
I'd really prefer you storing segments as `Entity<SegemntedEntitySegment>` rather than getting Owner through the component.